magento/pwa-studio
# Product Images Pre Fetching

## Description

Implemented Pre Fetching and Caching product images in the background by the SW.

This change speeds up the product page image retrieval by 7000% (500ms -> 2-7ms) on average by delegating the image fetch functionality to the Service Worker beforehand so the image is available when the component needs it. Also, this functionality is only implemented only when the network type is `4g`.

Note: Ideally I would only do this if the type is `WiFi` but the API for network details is not out yet. [It is still a draft.](https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API). Once it is out, we can change the `isFastNetwork` function of `networkUtils` to only work if `WiFi`.

## Related Issue
Closes #1769.
Closes #1766.

### Verification Steps
1. Build and Run the server using `yarn build && yarn stage:venia` command.
2. Go to the server URL in the browser.
3. Open the dev tools and go to the `Service Workers` sub-tab in the `Application` tab. Make sure the Service Worker status is `Activated and Running` if not refresh the page again.
4. Open any category and click on any product that has multiple images.
5. Open the `Network` tab in the dev tools and filter using `Img`. Clear the list for ease.
6. Click on any image thumbnail other than the first image. You should see that the time it takes to get the image should be significantly less than what the same steps in the develop branch take.

1. To verify Offline Capabilities, do steps 2 to 5 again in a new incognito tab.
2. Make the tab offline by changing from `Online` to `Offline` in the network tab.
3. Click on any image thumbnail other than the first image. Images should load unlike `develop` where they fail since network is not available.

## Screenshots / Screen Captures (if appropriate)

Speed Without Changes:

![Image Pre Fetching DEMO witPre Fetching DEMO without Changes](https://user-images.githubusercontent.com/35203638/67699472-33174480-f97a-11e9-8d9e-44dbd0961764.gif)

Speed With Changes:

![Image Pre Fetching DEMO with Changes](https://user-images.githubusercontent.com/35203638/67698967-4f66b180-f979-11e9-9248-a1ffc5cf9107.gif)

Offline Experience Without Changes:

![Image Pre Fetching Offline Demo without Changes](https://user-images.githubusercontent.com/35203638/67700056-39f28700-f97b-11e9-8532-e795615f017b.gif)

Offline Experience With Changes:

![Image Pre Fetching Offline Demo with Changes](https://user-images.githubusercontent.com/35203638/67704115-855c6380-f982-11e9-9131-e641d43135c0.gif)

## Checklist

* Verify that nothing has changed in the Development mode. These changes are for only `stage` mode.
* Merge only after #1871 and #1905 are merged.
